monkeytype/stubs.py

```diff
@@ -335,20 +335,29 @@ def render_annotation(anno: Any) -> str:
     if _is_optional(anno):
         elem_type = _get_optional_elem(anno)
         rendered = render_annotation(elem_type)
-        return 'Optional[' + rendered + ']'
+        rendered = 'Optional[' + rendered + ']'
     elif hasattr(anno, '__supertype__'):
-        return anno.__name__
+        rendered = anno.__name__
     elif getattr(anno, '__module__', None) == 'typing':
-        return repr(anno).replace('typing.', '')
-    elif anno is NoneType:
-        # This isn't my favorite thing in the world, but by convention NoneType
-        # and None are used interchangeably in signatures.
-        return 'None'
+        rendered = repr(anno).replace('typing.', '')
+    elif isinstance(anno, NoneType):
```

Comment:
The need to both keep this (for top-level `NoneType`) and add the search and replace below, reveals that there is a deeper bug hiding here, see #76. It's not just `None -> NoneType`, but every special case in this function is not applied within generic types, including e.g. `Union[X, NoneType]` -> `Optional[X]`.

Reply:
Agreed. If the nesting problem is handled, the search-replace hack would not be needed.